File: sdks/java/src/test/java/dev/atap/sdk/crypto/Base64UrlTest.java
package dev.atap.sdk.crypto;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class Base64UrlTest {

    @Test
    void testEncodeNoPadding() {
        byte[] data = "Hello, World!".getBytes();
        String encoded = Base64Url.encode(data);
        assertFalse(encoded.contains("="));
        assertFalse(encoded.contains("+"));
        assertFalse(encoded.contains("/"));
    }

    @Test
    void testRoundTrip() {
        byte[] original = "test data for base64url".getBytes();
        String encoded = Base64Url.encode(original);
        byte[] decoded = Base64Url.decode(encoded);
        assertArrayEquals(original, decoded);
    }

    @Test
    void testRoundTripBinaryData() {
        byte[] original = new byte[]{0, 1, 2, (byte) 255, (byte) 254, (byte) 253, 127, (byte) 128};
        String encoded = Base64Url.encode(original);
        byte[] decoded = Base64Url.decode(encoded);
        assertArrayEquals(original, decoded);
    }

    @Test
    void testEmptyInput() {
        byte[] empty = new byte[0];
        String encoded = Base64Url.encode(empty);
        assertEquals("", encoded);
        byte[] decoded = Base64Url.decode(encoded);
        assertEquals(0, decoded.length);
    }

    @Test
    void testEncodeUrlSafeCharacters() {
        // Bytes that would produce + and / in standard base64
        byte[] data = new byte[]{(byte) 0xFB, (byte) 0xFF, (byte) 0xFE};
        String encoded = Base64Url.encode(data);
        assertFalse(encoded.contains("+"));
        assertFalse(encoded.contains("/"));
    }

    @Test
    void testDecodeWithPadding() {
        // Should handle input that already has padding
        String withPadding = "SGVsbG8=";
        byte[] decoded = Base64Url.decode(withPadding);
        assertEquals("Hello", new String(decoded));
    }
}

File: sdks/php/tests/Crypto/Base64UrlTest.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Tests\Crypto;

use Atap\Sdk\Crypto\Base64Url;
use Atap\Sdk\Tests\TestCase;

class Base64UrlTest extends TestCase
{
    public function testEncodeEmpty(): void
    {
        $this->assertSame('', Base64Url::encode(''));
    }

    public function testDecodeEmpty(): void
    {
        $this->assertSame('', Base64Url::decode(''));
    }

    public function testRoundTrip(): void
    {
        $data = random_bytes(32);
        $encoded = Base64Url::encode($data);
        $decoded = Base64Url::decode($encoded);
        $this->assertSame($data, $decoded);
    }

    public function testNoPadding(): void
    {
        $encoded = Base64Url::encode('test');
        $this->assertStringNotContainsString('=', $encoded);
    }

    public function testUrlSafe(): void
    {
        // Use data that produces + and / in standard base64
        $data = "\xff\xff\xff";
        $encoded = Base64Url::encode($data);
        $this->assertStringNotContainsString('+', $encoded);
        $this->assertStringNotContainsString('/', $encoded);
    }

    public function testDecodeWithPadding(): void
    {
        // Standard base64 with padding
        $data = 'test';
        $standard = base64_encode($data);
        $urlSafe = rtrim(strtr($standard, '+/', '-_'), '=');
        $this->assertSame($data, Base64Url::decode($urlSafe));
    }

    public function testInvalidBase64Throws(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        Base64Url::decode('!!!invalid!!!');
    }

    public function testKnownValues(): void
    {
        // "Hello" in base64url should be "SGVsbG8"
        $this->assertSame('SGVsbG8', Base64Url::encode('Hello'));
        $this->assertSame('Hello', Base64Url::decode('SGVsbG8'));
    }

    public function testBinaryData(): void
    {
        $data = "\x00\x01\x02\x03\xfe\xff";
        $encoded = Base64Url::encode($data);
        $this->assertSame($data, Base64Url::decode($encoded));
    }
}
